src/translate.rs

```diff
@@ -21,7 +21,14 @@
 //! ```
 //!
 //! `ToGlibPtr`, `FromGlibPtrNone`, `FromGlibPtrFull` and `FromGlibPtrBorrow` work on `gpointer`s
-//! and support different modes of ownership transfer.
+//! and ensure correct ownership of values on the Rust side
+//! according to [Glib ownership transfer rules](https://gi.readthedocs.io/en/latest/annotations/giannotations.html).
+//!
+//! `FromGlibPtrNone` and `FromGlibPtrFull`
+//! must be called on values obtained from C,
+//! according to their `transfer` annotations.
+//! In particular, they turn floating references to strong ones,
```

Comment:
I'd move the mention of floating references into a separate paragraph with some more explanation what that is (maybe with link to the GObject docs) and why/when it matters. Also only `from_glib_none()` does this (on references that are still floating).

I wouldn't put floating references right at the top as for most users of the API this is just going to confuse and not actually relevant.

Reply:
> Also only from_glib_none() does this (on references that are still floating).

I wrote that because I understood that the contract on the Rust side ensured by the traits together is "no floating references on the Rust side". This section is less about what each function does, but more about what guarantees the user should expect and uphold.

I would also keep a mention of floating references here at the module level if possible, as this is the reason I dug myself into this hole :)

Reply:
Maybe a link to a section elsewhere that explains in more detail what that actually means?

> I wrote that because I understood that the contract on the Rust side ensured by the traits together is "no floating references on the Rust side".

Except for callbacks (see below), where we must not touch them. Whenever we get a floating reference in bindings that we can potentially take ownership of, we do so. To side-step all the problems that otherwise result from floating references (e.g. did `gtk_container_add()` steal my reference or not?)